src/main/java/hudson/plugins/copyartifact/CopyArtifact.java

```diff
@@ -324,24 +395,33 @@ public boolean perform(AbstractBuild<?,?> build, Launcher launcher, BuildListene
                     // Use subdir of targetDir with configuration name (like "jdk=java6u20")
                     ok |= perform(r, build, expandedFilter, expandedExcludes, targetDir.child(r.getParent().getName()),
                                   baseTargetDir, copier, console);
-                return ok;
+
+                if (!ok) {
+                    throw new CopyArtifactException(Messages.CopyArtifact_FailedToCopy(expandedProject, expandedFilter));
+                }
             } else {
-                return perform(src, build, expandedFilter, expandedExcludes, targetDir, baseTargetDir, copier, console);
+                if (!perform(src, build, expandedFilter, expandedExcludes, targetDir, baseTargetDir, copier, console)) {
+                    throw new CopyArtifactException(Messages.CopyArtifact_FailedToCopy(expandedProject, expandedFilter));
+                }
             }
         }
+        catch (CopyArtifactException cae) {
+            throw cae;
+        }
         catch (IOException ex) {
             Util.displayIOException(ex, listener);
-            ex.printStackTrace(listener.error(
-                    Messages.CopyArtifact_FailedToCopy(expandedProject, expandedFilter)));
-            return false;
+            String message = Messages.CopyArtifact_FailedToCopy(expandedProject, expandedFilter);
+            ex.printStackTrace(listener.error(message));
+            throw new CopyArtifactException(message);
```

Comment:
This whole `catch` block could simply be deleted, I think: let the `IOException` be thrown up to the caller and handled there.